File: lib/omnya_connector/origin_matcher.rb
require "uri"

module OmnyaConnector
  module OriginMatcher
    module_function

    def origin_allowed?(origin, allowed_origins)
      normalized_origin = normalize_origin(origin)
      return false if normalized_origin.nil?

      Array(allowed_origins).any? do |allowed_origin|
        origin_rule_matches?(normalized_origin, allowed_origin)
      end
    end

    def origin_rule_matches?(normalized_origin, allowed_origin)
      rule = allowed_origin.to_s.strip
      return false if rule.empty?

      if wildcard_rule?(rule)
        wildcard_rule_matches?(normalized_origin, rule)
      else
        normalize_origin(rule) == normalized_origin
      end
    end

    def wildcard_rule?(origin_rule)
      origin_rule.include?("*")
    end

    def wildcard_rule_matches?(normalized_origin, wildcard_rule)
      wildcard_match = wildcard_rule.match(/\Ahttps:\/\/\*\.([a-z0-9.-]+)(?::(\d+))?\z/i)
      return false unless wildcard_match

      wildcard_host = wildcard_match[1].downcase
      wildcard_port = wildcard_match[2]
      origin_uri = URI.parse(normalized_origin)
      origin_host = origin_uri.host.to_s.downcase

      # Wildcards only match subdomains, never the apex domain.
      return false unless origin_host.end_with?(".#{wildcard_host}")

      wildcard_port.nil? ? default_port?(origin_uri) : origin_uri.port.to_s == wildcard_port
    rescue URI::InvalidURIError
      false
    end

    def normalize_origin(origin)
      raw_origin = origin.to_s.strip
      return nil if raw_origin.empty?

      uri = URI.parse(raw_origin)
      return nil unless uri.scheme && uri.host

      scheme = uri.scheme.downcase
      host = uri.host.downcase

      standard_port = default_port?(uri)
      standard_port ? "#{scheme}://#{host}" : "#{scheme}://#{host}:#{uri.port}"
    rescue URI::InvalidURIError
      nil
    end

    def default_port?(uri)
      (uri.scheme == "https" && uri.port == 443) || (uri.scheme == "http" && uri.port == 80)
    end
    private_class_method :default_port?
  end
end

File: app/assets/javascripts/omnya_connector/controllers/origin_matcher.js
function defaultPortForScheme(scheme) {
  if (scheme === "https") return "443"
  if (scheme === "http") return "80"
  return null
}

function effectivePort(parsed) {
  if (parsed.port && parsed.port.length > 0) return parsed.port

  const scheme = parsed.protocol.replace(/:$/, "").toLowerCase()
  return defaultPortForScheme(scheme)
}

function defaultPort(parsed) {
  const scheme = parsed.protocol.replace(/:$/, "").toLowerCase()
  const port = effectivePort(parsed)
  const schemeDefaultPort = defaultPortForScheme(scheme)

  return schemeDefaultPort !== null && port === schemeDefaultPort
}

export function normalizeOrigin(origin) {
  const rawOrigin = String(origin || "").trim()
  if (rawOrigin.length === 0) return null

  let parsed
  try {
    parsed = new URL(rawOrigin)
  } catch (_error) {
    return null
  }

  if (!parsed.protocol || !parsed.hostname) return null

  const scheme = parsed.protocol.replace(/:$/, "").toLowerCase()
  const host = parsed.hostname.toLowerCase()
  const port = effectivePort(parsed)
  const includePort = port && !defaultPort(parsed)

  return includePort ? `${scheme}://${host}:${port}` : `${scheme}://${host}`
}

export function wildcardRuleMatches(normalizedOrigin, wildcardRule) {
  const wildcardMatch = String(wildcardRule || "").trim().match(/^https:\/\/\*\.([a-z0-9.-]+)(?::(\d+))?$/i)
  if (!wildcardMatch) return false

  let originUrl
  try {
    originUrl = new URL(normalizedOrigin)
  } catch (_error) {
    return false
  }

  const wildcardHost = wildcardMatch[1].toLowerCase()
  const wildcardPort = wildcardMatch[2]
  const originHost = (originUrl.hostname || "").toLowerCase()

  // Wildcards only match subdomains, never the apex domain.
  if (!originHost.endsWith(`.${wildcardHost}`)) return false

  if (!wildcardPort) return defaultPort(originUrl)

  const originPort = effectivePort(originUrl)
  return originPort === wildcardPort
}

export function wildcardRule(originRule) {
  return String(originRule || "").includes("*")
}

export function originRuleMatches(normalizedOrigin, allowedOrigin) {
  const rule = String(allowedOrigin || "").trim()
  if (rule.length === 0) return false

  if (wildcardRule(rule)) {
    return wildcardRuleMatches(normalizedOrigin, rule)
  }

  return normalizeOrigin(rule) === normalizedOrigin
}

export function originAllowed(origin, allowedOrigins) {
  const normalizedOrigin = normalizeOrigin(origin)
  if (!normalizedOrigin) return false

  return Array.from(allowedOrigins || []).some((allowedOrigin) => {
    return originRuleMatches(normalizedOrigin, allowedOrigin)
  })
}